akka-stream-tests/src/test/java/akka/stream/javadsl/TcpTest.java

```diff
@@ -165,50 +173,46 @@ public void mustReportClientConnectFailure() throws Throwable {
   }
 
   // compile only sample
-  public void constructSslContext() throws Exception {
-    ActorSystem system = null;
-
-    // #setting-up-ssl-context
-
-    // -- setup logic ---
-
-    AkkaSSLConfig sslConfig = AkkaSSLConfig.get(system);
-
-    // Don't hardcode your password in actual code
-    char[] password = "abcdef".toCharArray();
+  // #setting-up-ssl-engine
+  public SSLEngine createSSLEngine(TLSRole role) {
+    try {
+      // Don't hardcode your password in actual code
+      char[] password = "abcdef".toCharArray();
 
-    // trust store and keys in one keystore
-    KeyStore keyStore = KeyStore.getInstance("PKCS12");
-    keyStore.load(getClass().getResourceAsStream("/tcp-spec-keystore.p12"), password);
+      // trust store and keys in one keystore
+      KeyStore keyStore = KeyStore.getInstance("PKCS12");
+      keyStore.load(getClass().getResourceAsStream("/tcp-spec-keystore.p12"), password);
 
-    TrustManagerFactory tmf = TrustManagerFactory.getInstance("SunX509");
-    tmf.init(keyStore);
+      TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance("SunX509");
+      trustManagerFactory.init(keyStore);
 
-    KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance("SunX509");
-    keyManagerFactory.init(keyStore, password);
+      KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance("SunX509");
+      keyManagerFactory.init(keyStore, password);
 
-    // initial ssl context
-    SSLContext sslContext = SSLContext.getInstance("TLS");
-    sslContext.init(keyManagerFactory.getKeyManagers(), tmf.getTrustManagers(), new SecureRandom());
+      // initial ssl context
+      SSLContext sslContext = SSLContext.getInstance("TLSv1.2");
+      sslContext.init(
+          keyManagerFactory.getKeyManagers(),
+          trustManagerFactory.getTrustManagers(),
+          new SecureRandom());
 
-    // protocols
-    SSLParameters defaultParams = sslContext.getDefaultSSLParameters();
-    String[] defaultProtocols = defaultParams.getProtocols();
-    String[] protocols = sslConfig.configureProtocols(defaultProtocols, sslConfig.config());
-    defaultParams.setProtocols(protocols);
+      SSLEngine engine = sslContext.createSSLEngine();
```

Comment:
That's not a recommended way to show in the documentation. The `sslContext` should only be created once and not multiple times. First to avoid creating multiple SSLContexts because of overhead but more importantly because the SSLContext contains some global caches necessary to make repeated connections between TLS peers more efficient (session resumption and session tickets).

Reply:
Could we provide a utility method for users

```scala
def engineFromContext(ctx: SSLContext, role: TLSRole): () => SSLEngine
```

Reply:
I thought we should stay away from utilities here because where do we stop, and suddenly we have a new sslconfig?

At higher layers we make it more convenient, such as the config based in Artery and I guess there is something in Akka HTTP as well.

I'll update the example and make it clear that SSLContext should only be created once.

Reply:
updated doc example in 87c938ad78fd36fe51d8ccbb2f83f08881c823f0